pkg/client/client_venconn.go

```diff
@@ -0,0 +1,176 @@
+// no authentication.
+func NewVenConnClient(c *http.Client, agentMetadata *api.AgentMetadata, baseURL, kubeconfigPath, installNamespace, venConnName string) (*VenConnClient, error) {
+	cfg, err := loadRESTConfig(kubeconfigPath)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	cfg.Impersonate = rest.ImpersonationConfig{
+		UserName: fmt.Sprintf("system:serviceaccount:%s:venafi-connection", installNamespace),
+	}
+
+	clientset, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	clientset.DiscoveryClient.RESTClient()
+
+	restMapper, err := apiutil.NewDynamicRESTMapper(cfg, &http.Client{})
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	// This Kubernetes client only needs to be able to read and write the
+	// VenafiConnection resources and read Secret resources.
+	scheme := runtime.NewScheme()
+	_ = venapi.AddToScheme(scheme)
+	_ = corev1.AddToScheme(scheme)
+
+	handler, err := venafi_client.NewConnectionHandler(
+		"venafi-kubernetes-agent/"+version.PreflightVersion,
+		"venafi-kubernetes-agent.jetstack.io",
+		"VenafiKubernetesAgent",
+		cfg,
+		scheme,
+		restMapper,
+		nil,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	go func() {
+		err := handler.CacheRunnable().Start(context.TODO())
+		if err != nil {
+			panic(err)
+		}
+	}()
+
+	return &VenConnClient{
+		baseURL:          baseURL,
+		agentMetadata:    agentMetadata,
+		connHandler:      handler,
+		installNamespace: installNamespace,
+		venConnName:      venConnName,
+		client:           c,
+	}, nil
+}
+
+func (c *VenConnClient) PostDataReadingsWithOptions(readings []*api.DataReading, opts Options) error {
+	return c.PostDataReadings(opts.OrgID, opts.ClusterID, readings)
+}
+
+// PostDataReadings uploads the slice of api.DataReading to the Jetstack Secure backend to be processed for later
+// viewing in the user-interface.
+func (c *VenConnClient) PostDataReadings(orgID, clusterID string, readings []*api.DataReading) error {
+	payload := api.DataReadingsPost{
+		AgentMetadata:  c.agentMetadata,
+		DataGatherTime: time.Now().UTC(),
+		DataReadings:   readings,
+	}
+	data, err := json.Marshal(payload)
+	if err != nil {
+		return err
+	}
+
+	res, err := c.Post(filepath.Join("/api/v1/org", orgID, "datareadings", clusterID), bytes.NewBuffer(data))
+	if err != nil {
+		return err
+	}
+	defer res.Body.Close()
+
+	if code := res.StatusCode; code < 200 || code >= 300 {
+		errorContent := ""
+		body, err := ioutil.ReadAll(res.Body)
+		if err == nil {
+			errorContent = string(body)
+		}
+
+		return fmt.Errorf("received response with status code %d. Body: [%s]", code, errorContent)
+	}
+
+	return nil
+}
+
+// Post performs an HTTP POST request.
+func (c *VenConnClient) Post(path string, body io.Reader) (*http.Response, error) {
+	_, token, err := c.connHandler.Get(context.Background(), c.installNamespace, auth.Scope{}, types.NamespacedName{Name: c.venConnName, Namespace: c.installNamespace})
```

Comment:
I don't really understand what `requestorNamespace` means in the context of `Get`. There is pretty much no documentation around this argument: https://github.com/jetstack/venafi-connection-lib/blob/fec6e292/venafi_client/handler.go#L245-L249

Reply:
The `requestorNamespace` is used to validate cross-namespace VenafiConnection references.
It is the namespace of the resource referencing the VenafiConnection.

Reply:
In the case of the Venafi Agent, I assume this will just be the namespace in which the agent is installed.

Reply:
Thanks. I added a comment on top of the Get func to explain what `requestorNamespace` is about.